java/monkey/src/main/java/Lexer.java

```diff
@@ -20,33 +18,58 @@ public Token nextToken() {
             case ',' -> TokenType.COMMA.token();
             case ';' -> TokenType.SEMI.token();
             case '+' -> TokenType.PLUS.token();
-            case '=' -> TokenType.EQUAL.token();
+            case '=' -> {
```

Comment:
I'm thinking we swap this to
```java
case '=' when peekNext() == '=' ->  {this.advance(); yield TokenType.EQUAL.token();}
case '=' -> TokenType.ASSIGN.token();
case '!' when peekNext() == '=' -> {this.advance(); yield TokenType.NOT_EQUAL.token();}
case '!' -> TokenType.BANG.token();
```

Reply:
I sadly don't think

```java
case '=' when peekNext() == '=' ->  {this.advance(); yield TokenType.EQUAL.token();}
``` 

is valid syntax. Patter matching can only be done when capturing the value like:

```java
case Character c when c == '=' && peekNext() == '=' ->  {this.advance(); yield TokenType.EQUAL.token();}
```

and for that to work all branches would have to begin with `case Character c where c == '...'` and I believe that would be too verbose and less clear then an if in the special cases